src/shogun/classifier/AveragedPerceptron.cpp

```diff
@@ -14,105 +15,99 @@
 
 using namespace shogun;
 
-CAveragedPerceptron::CAveragedPerceptron()
-: CLinearMachine()
+CAveragedPerceptron::CAveragedPerceptron() :
+ 	CAveragedPerceptron(nullptr, nullptr, 1000, 0.1)
 {
-	init();
-
 }
 
-CAveragedPerceptron::CAveragedPerceptron(CDotFeatures* traindat, CLabels* trainlab)
-: CLinearMachine()
+CAveragedPerceptron::CAveragedPerceptron(CDotFeatures* traindat, CLabels* trainlab) :
+	CAveragedPerceptron(traindat, trainlab, 1000, 0.1)
 {
-	set_features(traindat);
-	set_labels(trainlab);
-	init();
 }
 
-CAveragedPerceptron::~CAveragedPerceptron()
+CAveragedPerceptron::CAveragedPerceptron(
+	CDotFeatures* traindat,
+	CLabels* trainlab,
+	int32_t max_iters,
+	float64_t learning_rate
+) : CIterativeMachine<CLinearMachine>()
 {
+	if (trainlab) set_labels(trainlab);
+	if (traindat) init_model((CFeatures*) traindat);
+	
+	set_max_iterations(max_iters);
+	set_learn_rate(learning_rate);
+	
+	SG_ADD(&learn_rate, "learn_rate", "Learning rate.", ParameterProperties::HYPER);
+	SG_ADD(&tmp_w, "tmp_w", "Sum of weights across iterations.");
+	SG_ADD(&tmp_bias, "tmp_bias", "Sum of biases across iterations.");
 }
 
-void CAveragedPerceptron::init()
+CAveragedPerceptron::~CAveragedPerceptron()
 {
-	max_iter = 1000;
-	learn_rate = 0.1;
-	SG_ADD(&max_iter, "max_iter", "Maximum number of iterations.", ParameterProperties::HYPER);
-	SG_ADD(&learn_rate, "learn_rate", "Learning rate.", ParameterProperties::HYPER);
 }
 
-bool CAveragedPerceptron::train_machine(CFeatures* data)
+void CAveragedPerceptron::init_model(CFeatures* data)
 {
-	ASSERT(m_labels)
-
 	if (data)
 	{
 		if (!data->has_property(FP_DOT))
 			SG_ERROR("Specified features are not of type CDotFeatures\n")
 		set_features((CDotFeatures*) data);
-	}
-	ASSERT(features)
-	bool converged=false;
-	int32_t iter=0;
-	SGVector<int32_t> train_labels = binary_labels(m_labels)->get_int_labels();
-	int32_t num_feat=features->get_dim_feature_space();
-	int32_t num_vec=features->get_num_vectors();
 
-	ASSERT(num_vec==train_labels.vlen)
+		SG_REF(data);
+		SG_UNREF(m_continue_features);
+		m_continue_features = data->as<CDotFeatures>();
```

Comment:
@shubham808 is this necessary in subclasses?

Reply:
yeah we should do this in each subclass

Reply:
I did the required changes, and brought back the assertions. Averaged perceptron is pretty much the same as normal perceptron (except for averaging), so I imitated what was done in CPerceptron class. I think it can benefit from the changes you mentioned (using linalg and putting assertions), I can do them after this request.

Reply:
I don’t think it’s good as then we’d have the same code in every subclass. Plus if someone forgets to do this.... can’t this be something that is done automatically?

Reply:
the problem is we wont know what to cast data to as it is only known in subclass so doing this here makes more sense

Reply:
But we had this discussion before no?: there is no reason why the type of continue features is DotFestures. It should be just Features. The reference should be stored automatically in the base class, and the subclasses can check/assert the correct type at runtime in init_model (but not store them here, just check) and then the iteration method can do a static cast as that’s safe then. This copy pasting of ref/unref code is not good, error prone.
See what I mean?

Reply:
Maybe I’m confused about this. Need to check he code my computer ... will do soon. But pls also see if the above makes sense

Reply:
Just checked. The continue features are of type Features. This means that all that needs to happen here is the runtime assertion (as is fine, though there also is the enum, which is faster) that the features have the correct type. The code for storing the reference should be in IterativeMachine. Might need some refactoring for the perceptron as well ...

Reply:
No need for the as cast btw here as the property is already asserted above